verilog/analysis/checkers/mismatched_labels_rule.cc

```diff
@@ -0,0 +1,119 @@
+// Copyright 2017-2019 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/mismatched_labels_rule.h"
+
+#include <string>
+#include <vector>
+#include <set>
+
+#include "absl/strings/str_cat.h"
+#include "common/analysis/citation.h"
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/matcher/bound_symbol_manager.h"
+#include "common/analysis/matcher/matcher.h"
+#include "common/analysis/syntax_tree_search.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/analysis/descriptions.h"
+#include "verilog/analysis/lint_rule_registry.h"
+#include "verilog/CST/identifier.h"
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+
+// Register the lint rule
+VERILOG_REGISTER_LINT_RULE(MismatchedLabelsRule);
+
+absl::string_view MismatchedLabelsRule::Name() { return "mismatched-labels"; }
+const char MismatchedLabelsRule::kTopic[] = "mismatched-labels";
+const char MismatchedLabelsRule::kMessageMismatch[] =
+    "Begin/end block labels must match.";
+const char MismatchedLabelsRule::kMessageMissing[] =
+    "Matching begin label is missing.";
+
+std::string MismatchedLabelsRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat(
+      "Labels mismatch. See:", GetStyleGuideCitation(kTopic), ".");
+}
+
+void MismatchedLabelsRule::HandleSymbol(
+    const verible::Symbol& symbol, const verible::SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+
+  if (matcher_.Matches(symbol, &manager)) {
+
+    // Find the corresponding kEnd node.
+    // Note that this may find more than one kEnd node if the blocks
+    // are nested, in this case using the very last result ensures
+    // we use the kEnd corresponding to the given kBegin
+    std::vector<verible::TreeSearchMatch> end =
+      SearchSyntaxTree(context.top(), NodekEnd());
+
+    // This should never really happen with correct syntax
+    if (end.size() == 0) {
+      return;
+    }
+
+    // Find the label of the end statement
+    std::vector<verible::TreeSearchMatch> end_label_node =
+      SearchSyntaxTree(*end[end.size() - 1].match, NodekLabel());
```

Comment:
Could this search be simplified to a direct access, given the way it is constructed in verilog.y?  Could also be a re-usable CST function, like `GetEndLabel()` (may return nullptr).

Reply:
Sure, that makes sense, thanks! I implemented this as `GetEndLabelTokenInfo` as I need both the actual label as well as its TokenInfo for the linter rule implementation